TDI-37709 Date/Time issues in DI conversion

francisu opened the pull request:
**Please check if the PR fulfills these requirements**

- [X] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [X] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [X] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)


**What kind of change does this PR introduce?**

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

https://jira.talendforge.org/browse/TDI-37709


**What is the new behavior?**

Fixed.


**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**: 

Goes with corresponding change in tcommon-studio-se

build-talend commented:
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/316/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/316/)

RyanSkraba commented on daikon/src/main/java/org/talend/daikon/di/DiOutgoingSchemaEnforcer.java:
```diff
@@ -12,19 +12,21 @@
 import org.talend.daikon.avro.SchemaConstants;
 import org.talend.daikon.avro.converter.IndexedRecordConverter.UnmodifiableAdapterException;
 
+import java.math.BigDecimal;
```
https://github.com/Talend/tools/tree/master/tools-java-formatter says `java;javax;org;com;` ... doesn't need to be changed here for this PR, but it would be great if we stopped flipping imports!

francisu replied:
I formatting according to the Maven rules (did a mvn install -Dformat). Daikon is picky about the formatting.

gonchar-ivan replied:
Does -Dformat also organize imports?

francisu replied:
It does not, I fixed the code formatting.

RyanSkraba commented on daikon/src/main/java/org/talend/daikon/di/DiOutgoingSchemaEnforcer.java:
```diff
@@ -171,11 +173,11 @@ public void put(int i, Object v) {
     /**
      * {@inheritDoc}
-     * 
+     * <p>
```
It looks like you might have updated intellij to 2016.3 :)  I lost all of my code formatting setup too and had to re-apply the Talend settings.  It would be great if you changed it back (again, not a blocking change for this PR).

francisu replied:
Fixed, thanks!

RyanSkraba commented on daikon/src/main/java/org/talend/daikon/di/DiOutgoingSchemaEnforcer.java:
```diff
@@ -187,19 +189,28 @@ public Object get(int pojoIndex) {
+        if (logicalType != null) {
+            if (logicalType == LogicalTypes.date() || logicalType == LogicalTypes.timeMillis()
+                    || logicalType == LogicalTypes.timestampMillis()) {
+                return new Date(value instanceof Integer ? (Integer) value : (Long) value);
```
This might be faulty!  https://avro.apache.org/docs/1.8.0/spec.html#Logical+Types

If it is LogicalTypes.date(), it is the number of days since the unix epoch (1970-01-01), not the number of milliseconds like the constructor for Date().

You probably want to use something like:

```
// TODO: replace this with java.time when java 8 is available.
Calendar c = Calendar.getInstance();
c.setTime(0L);
c.add(Calendar.DATE, (Integer) value);  
return c.getTime()
```

What do you think about adding `timeMicros` and `timestampMicros` at the same time?  Just divide the long by 1000.  We lose precision because Date doesn't support microseconds... but java.time does in java 8... one day :)

gonchar-ivan replied:
The only thing from my side here - transformValue() method is critical for performance. It is called many times, so we need to make it as efficient as possible.

gonchar-ivan replied:
Are you sure we can map timeMillis logical type to a java Date?

timeMillis represents the number of milliseconds after midnight. It is not related to particular calendar, time zone or date

gonchar-ivan replied:
I need to explain one thing here.

valueField parameter in transformValue() method is a Field of design schema.

Design schema comes from Studio. It comes from SchemaEditor.
As I understand, when user modifies something in SchemaEditor, Studio handles MetadataTable object.
Then it converts it to avro (design) schema. 
Date field in design schema won't contain logicalType property. It will be AvroUtils._date().

Maybe you need additionally runtime field parameter (comes from IndexedRecord emitted by component) in this method

RyanSkraba commented on daikon/src/test/java/org/talend/daikon/di/DiOutgoingSchemaEnforcerTest.java:
```diff
@@ -38,37 +41,62 @@
      */
     private static IndexedRecord record;
 
+    private static void setLogicalType(Schema schema, LogicalType lt) throws Exception {
```
This is unnecessary (see below).

RyanSkraba commented on daikon/src/test/java/org/talend/daikon/di/DiOutgoingSchemaEnforcerTest.java:
```diff
@@ -38,37 +41,62 @@
                 .name("createdDate").prop(DiSchemaConstants.TALEND6_COLUMN_TALEND_TYPE, "id_Date") //
                 .prop(DiSchemaConstants.TALEND6_COLUMN_PATTERN, "yyyy-MM-dd'T'HH:mm:ss'000Z'").type().nullable().longType() //
                 .noDefault() //
+                .name("logicalDate").prop("logicalType", "date").type(Schema.create(Schema.Type.INT)).noDefault() //
```
There's a better way to specify logical types with a SchemaBuilder.

Try 
```
.name("logicalDate").type(LogicalTypes.date().addToSchema(Schema.create(Schema.Type.INT))).noDefault() //
.name("logicalTime").type(LogicalTypes.timeMillis().addToSchema(Schema.create(Schema.Type.INT))).noDefault() //
.name("logicalTimestamp").type(LogicalTypes.timestampMillis().addToSchema(Schema.create(Schema.Type.LONG))).noDefault() //
```

(You can get rid of the `setLogicalType()` method and the `throws Exception` here).

francisu replied:
Thank you *so much* I spent a lot of time trying to figure out how to build a logical type with the SchemaBuilder; this is just what I needed.

RyanSkraba commented on daikon/src/test/java/org/talend/daikon/di/DiOutgoingSchemaEnforcerTest.java:
```diff
@@ -99,6 +127,9 @@ public void testGetByIndex() {
         assertThat(enforcer.get(2), equalTo((Object) 100));
         assertThat(enforcer.get(3), equalTo((Object) true));
         assertThat(enforcer.get(4), equalTo((Object) new Date(1467170137872L)));
+        assertThat(enforcer.get(5), equalTo((Object) new Date(1000L)));
```
The expected result is a thousand days after 1970-01-01 for column 5.

francisu replied:
Agreed.

pbabin commented:
Worth to note that TDS logical types DATE, TIME and TIMESTAMP (compliant with Avro) are respectively mapped in Studio to Date, int and long

francisu commented:
@pbabin, we need a standardized mapping of the Avro types to DI. I'm not sure what to do here. I would argue (along with the Snowflake people) that TIME and TIMESTAMP are more naturally mapped to Date so that you can get the date manipulation functions. Why did you chose the mappings you chose?

build-talend commented:
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/329/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/329/)

francisu commented:
I think I addressed all of the issues in my changes, and checked this out with the Snowflake tests that I have. @gonchar-ivan @RyanSkraba can you have another look.

Ivan - I did not want to use the DI encoding information in transformValue() as I think checking for the Avro logical type is more clear.

I did revise the mapping as you suggest that a Avro logical time is mapped to an integer instead of a date. The others remain mapped to dates.

build-talend commented:
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/331/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/331/)

gonchar-ivan commented on daikon/pom.xml:
```diff
@@ -24,6 +24,11 @@
 	</properties>
 	<dependencies>
 		<dependency>
+			<groupId>joda-time</groupId>
```
We need to align joda-time version upon all projects we develop (in components/daikon).
I'm checking how to do it better

gonchar-ivan replied:
2.8.2 version is ok here. The same version is used everywhere:
1. in components-api
2. it is also packed in default maven repo, which is provided with Studio, so it is available for job processes.

What I want to do here is put joda-time version dependencyManagement block in daikon-parent.
(Also to import dependencies from dependencyManagement of daikon-parent to components parent)
But it will be done in separate Jira issue

gonchar-ivan commented on daikon/src/main/java/org/talend/daikon/avro/AvroUtils.java:
```diff
@@ -5,14 +5,11 @@
-import org.talend.daikon.avro.converter.ConvertCharacter;
-import org.talend.daikon.avro.converter.ConvertDate;
-import org.talend.daikon.avro.converter.ConvertShort;
+import org.talend.daikon.avro.converter.*;
```
Shouldn't we unfold asterisk imports? 
What convention do we have about it?

gonchar-ivan commented on daikon/src/main/java/org/talend/daikon/avro/AvroUtils.java:
```diff
@@ -41,10 +38,23 @@ public static Schema _character() {
 
     private static final ConvertDate DATE_TYPE = new ConvertDate();
 
+    // FIXME - remove this one, this is not the date representation we ultimately want to use
```
I suggest to create Jira tickets for TODOs and FIXMEs.
We always forget about FIXME

gonchar-ivan commented on daikon/src/main/java/org/talend/daikon/di/DiOutgoingSchemaEnforcer.java:
```diff
@@ -12,19 +12,22 @@
 import org.talend.daikon.avro.SchemaConstants;
 import org.talend.daikon.avro.converter.IndexedRecordConverter.UnmodifiableAdapterException;
 
+import static org.talend.daikon.di.DiSchemaConstants.TALEND6_COLUMN_TALEND_TYPE;
```
Eclipse puts static imports above.
Seems, IDEA vice versa (((

build-talend commented:
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/342/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/342/)

build-talend commented:
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/343/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/343/)

sgandon commented:
Please be carefull when you change daikon depenencies this, PR breaks the Components builds because of joda time add a new import to the bundle.
for oinstance: https://ci-common.datapwn.com/job/components_pull-request_master/995/testReport/junit/org.talend.components.api/ComponentsPaxExamOptionsTest/test/
